pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+		var jobState model.JobState
+
+		dataJob := bktJobs.Get(key)
+		err = json.Unmarshal(dataJob, &job)
+		if err != nil {
+			return nil, err
+		}
+
+		dataState := bktState.Get(key)
+		err = json.Unmarshal(dataState, &jobState)
+		if err != nil {
+			return nil, err
+		}
+
+		info := model.JobWithInfo{
+			Job:   job,
+			State: jobState,
+		}
+		infos = append(infos, info)
+	}
+
+	return infos, nil
+}
+
+// GetJobHistory returns the job (and execution) history for the provided options
+func (b *BoltJobStore) GetJobHistory(ctx context.Context,
+	jobID string,
+	options jobstore.JobHistoryFilterOptions) ([]model.JobHistory, error) {
+	var history []model.JobHistory
+	err := b.database.View(func(tx *bolt.Tx) (err error) {
+		history, err = b.getJobHistory(tx, jobID, options)
+		return
+	})
+
+	return history, err
+}
+
+func (b *BoltJobStore) getJobHistory(tx *bolt.Tx, jobID string,
+	options jobstore.JobHistoryFilterOptions) ([]model.JobHistory, error) {
+	var history []model.JobHistory
+
+	if !options.ExcludeJobLevel {
+		if bkt, err := NewBucketPath(BucketPathJobHistory, jobID).Get(tx, false); err != nil {
+			return nil, err
+		} else {
+			err = bkt.ForEach(func(key []byte, data []byte) error {
+				var item model.JobHistory
+
+				err := json.Unmarshal(data, &item)
+				if err != nil {
+					return err
+				}
+
+				history = append(history, item)
+				return nil
+			})
+
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if !options.ExcludeExecutionLevel {
+		// 	// Get the executions for this JobID
+		if bkt, err := NewBucketPath(BucketPathExecutionHistory, jobID).Get(tx, false); err != nil {
+			return nil, err
+		} else {
+			err = bkt.ForEach(func(key []byte, data []byte) error {
+				var item model.JobHistory
+
+				err := json.Unmarshal(data, &item)
+				if err != nil {
+					return err
+				}
+
+				history = append(history, item)
+				return nil
+			})
+
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Filter out anything before the specified Since time
+	history = lo.Filter(history, func(item model.JobHistory, index int) bool {
+		return item.Time.Unix() >= options.Since
+	})
+
+	sort.Slice(history, func(i, j int) bool { return history[i].Time.UTC().Before(history[j].Time.UTC()) })
+
+	return history, nil
+}
+
+// CreateJob creates a new record of a job in the data store
+func (b *BoltJobStore) CreateJob(ctx context.Context, job model.Job) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.createJob(tx, job)
+	})
+}
+
+func (b *BoltJobStore) createJob(tx *bolt.Tx, job model.Job) error {
+	if b.jobExists(tx, job.ID()) {
+		return jobstore.NewErrJobAlreadyExists(job.Metadata.ID)
+	}
+
+	jobIDKey := []byte(job.Metadata.ID)
+
+	jobState := model.JobState{
+		JobID:      job.Metadata.ID,
+		State:      model.JobStateNew,
+		Version:    1,
+		CreateTime: b.clock.Now().UTC(),
+		UpdateTime: b.clock.Now().UTC(),
+	}
+	data, err := json.Marshal(jobState)
+	if err != nil {
+		return err
+	}
```

Comment:
as discussed earlier, adding an envelope around each payload before it is written to the data store can improve our versioning, backward compatibility and resiliency to future changes, such as changing the serialization from json to a binary format